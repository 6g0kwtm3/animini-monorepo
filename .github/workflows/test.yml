name: Test

on:
  push:
    branches:
      - master
  pull_request:
    branches:
      - master

permissions:
  contents: read

jobs:
  test:
    name: Test E2E
    runs-on: ubuntu-latest
    container:
      image: mcr.microsoft.com/playwright:v1.43.1-jammy

    steps:
      - uses: actions/checkout@v3
      - uses: pnpm/action-setup@v2
      - uses: actions/setup-node@v3
        with:
          node-version: 20
          cache: "pnpm"

      - name: Install dependencies ðŸ“¦
        run: pnpm i --frozen-lockfile

      - name: Await CF Pages
        uses: WalshyDev/cf-pages-await@v1
        id: cf-pages
        with:
          apiToken: ${{ secrets.CF_API_TOKEN  }}
          accountId: 'e9b0348e83d33e46fbb82c3271317d70'
          project: 'animini-pwa'

      - name: Test E2E
        run: pnpm run test
        env:
          ANILIST_TEST_TOKEN: ${{ secrets.ANILIST_TEST_TOKEN }}
          BASE_URL: ${{ steps.cf-pages.outputs.url }}
          HOME: /root

  typecheck:
    runs-on: ubuntu-latest

    name: Typecheck
    steps:
      - uses: actions/checkout@v3
      - uses: pnpm/action-setup@v2
      - uses: actions/setup-node@v3
        with:
          node-version: 20
          cache: "pnpm"

      - name: Install dependencies ðŸ“¦
        run: pnpm i --frozen-lockfile

      - run: pnpm run paraglide

      - name: Typecheck
        run: pnpm run check

  format:
    runs-on: ubuntu-latest
    name: Lint & Format
    permissions:
      contents: write

    steps:
      - uses: actions/checkout@v3
        with:
          token: ${{ secrets.PAT }}
      - uses: pnpm/action-setup@v2
      - uses: actions/setup-node@v3
        with:
          node-version: 20
          cache: "pnpm"

      - name: Install dependencies ðŸ“¦
        run: pnpm i --frozen-lockfile

      - name: Lint
        run: pnpm run lint --fix

      - name: Format
        run: pnpm run format

      - name: Commit changes
        uses: stefanzweifel/git-auto-commit-action@v4
        with:
          commit_message: "refactor: run format"
