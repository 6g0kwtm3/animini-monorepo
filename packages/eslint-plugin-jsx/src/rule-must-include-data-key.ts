import {
	type TSESTree,
	AST_NODE_TYPES,
} from "@typescript-eslint/typescript-estree"
import { ASTUtils } from "@typescript-eslint/utils"
import type {
	RuleContext,
	RuleModule,
} from "@typescript-eslint/utils/ts-eslint"

export const rule: RuleModule<string> = {
	meta: {
		type: "suggestion",
		docs: { description: "" },
		schema: [],
		messages: {
			"data-key-must-match-key":
				"JSX element with 'key' prop must also have a matching 'data-key' prop.",
		},
		fixable: "code",
	},
	defaultOptions: [],
	create(context: RuleContext<string, []>) {
		return {
			JSXOpeningElement(node: TSESTree.JSXOpeningElement) {
				const byName = Object.fromEntries(
					node.attributes.flatMap(
						(attribute): [string, TSESTree.JSXAttribute][] => {
							switch (attribute.type) {
								case AST_NODE_TYPES.JSXAttribute:
									return [
										[
											(attribute.name.type === AST_NODE_TYPES.JSXIdentifier
												? attribute.name
												: attribute.name.name
											).name,
											attribute,
										],
									]
								case AST_NODE_TYPES.JSXSpreadAttribute:
									return []
								default:
									attribute satisfies never
									throw new Error("unreachable")
							}
						}
					)
				)

				const key = byName.key

				if (!key) {
					return
				}

				const dataKey = byName["data-key"]

				if (!dataKey || !isEqual(context, key, dataKey)) {
					context.report({
						node: key,
						messageId: "data-key-must-match-key",
						data: {},
						fix(fixer) {
							const fixes = []

							if (dataKey) {
								const fix = fixer.remove(dataKey)
								fix.range = [fix.range[0] - 1, fix.range[1]]
								fixes.unshift(fix)
							}
							return [
								...fixes,
								fixer.insertTextAfter(
									key,
									` data-key${key.value ? `=${context.sourceCode.getText(key.value)}` : ""}`
								),
							]
						},
					})
				}
			},
		}
	},
}

function isEqual(
	context: RuleContext<string, []>,
	a: TSESTree.JSXAttribute,
	b: TSESTree.JSXAttribute
) {
	if (a === b) {
		return true
	}

	if (a.value === b.value) {
		return true
	}

	if (a.value == null || b.value == null) {
		return false
	}

	if (
		a.value.type === AST_NODE_TYPES.JSXElement
		|| b.value.type === AST_NODE_TYPES.JSXElement
	) {
		return true
	}

	const aExpression = "expression" in a.value ? a.value.expression : a.value

	const evaluatedA = getStaticValue(aExpression, context)
	const bExpression = "expression" in b.value ? b.value.expression : b.value

	const evaluatedB = getStaticValue(bExpression, context)

	if (evaluatedA && evaluatedB) {
		return evaluatedA.value === evaluatedB.value
	}

	return stableStringify(context, a.value) === stableStringify(context, b.value)
}

function stableStringify(
	context: RuleContext<string, []>,
	obj: TSESTree.JSXExpression | TSESTree.Literal
) {
	return JSON.stringify(obj, (key, value: unknown) => {
		if (key === "loc" || key === "range" || key === "parent") {
			return undefined
		}

		if (typeof value !== "object") {
			return value
		}

		if (value == null) {
			return value
		}

		if (Array.isArray(value)) {
			return value
		}

		return (
			getStaticValue(value as TSESTree.Node, context)
			?? Object.fromEntries(
				Object.entries(value).sort(([a], [b]) => a.localeCompare(b))
			)
		)
	})
}

function getStaticValue(node: TSESTree.Node, context: RuleContext<string, []>) {
	return ASTUtils.getStaticValue(node, context.sourceCode.getScope(node))
}
