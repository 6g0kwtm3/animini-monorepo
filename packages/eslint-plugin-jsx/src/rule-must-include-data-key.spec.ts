import * as typescriptParser from "@typescript-eslint/parser"
import { createRuleTester } from "eslint-vitest-rule-tester"
import { describe } from "node:test"
import { expect, test } from "vitest"
import { rule } from "./rule-must-include-data-key"

void describe("rule-must-include-data-key", () => {
	const { valid, invalid } = createRuleTester({
		rule,
		languageOptions: {
			ecmaVersion: 6,
			parser: typescriptParser,
			parserOptions: { ecmaFeatures: { jsx: true } },
		},
	})

	test("allows equal key and data-key", async () => {
		await valid(
			`
			<li key="foo" data-key="foo" />
			`
		)
	})

	test("allows equal key and data-key with boolean value", async () => {
		await valid(
			`
			<li key data-key />
			`
		)
	})

	test("allows equal key and data-key with computed value", async () => {
		await valid(
			`
      <li key={foo + bar} data-key={foo + bar} />
      `
		)
	})

	test("ignores whitespace", async () => {
		await valid(
			`
      <li key={foo + bar} data-key={foo +\nbar} />
      `
		)
	})

	test("substitutes vars", async () => {
		await valid(
			`
			const bar = "bar";
      <li key="bar" data-key={bar} />
      `
		)
	})

	test("allows empty key and data-key", async () => {
		await valid(
			`
      <li />
      `
		)
	})

	test("reports error for missing data-key", async () => {
		const { result } = await invalid({
			code: `
        <li key="foo" />
        `,
			errors: ["data-key-must-match-key"],
		})
		expect(result.output).toMatchSnapshot()
	})

	test("reports error for missing data-key with boolean value", async () => {
		const { result } = await invalid({
			code: `
        <li key data-key="foo" />
        `,
			errors: ["data-key-must-match-key"],
		})
		expect(result.output).toMatchSnapshot()
	})

	test("reports error for different data-key", async () => {
		const { result } = await invalid({
			code: `
        <li key={foo + bar} />
        `,
			errors: ["data-key-must-match-key"],
		})
		expect(result.output).toMatchSnapshot()
	})

	test("reports error for different data-key with computed value", async () => {
		const { result } = await invalid({
			code: `
				<li key="foo" data-key={foo} />
        `,
			errors: ["data-key-must-match-key"],
		})
		expect(result.output).toMatchSnapshot()
	})

	test("reports error when data-key is a boolean attribute without value", async () => {
		const { result } = await invalid({
			code: `
				<li key="foo" data-key />
        `,
			errors: ["data-key-must-match-key"],
		})
		expect(result.output).toMatchSnapshot()
	})

	test("reports error for missing data-key with computed value", async () => {
		const { result } = await invalid({
			code: `
        <li key={foo + bar} data-key="foo" />
        `,
			errors: ["data-key-must-match-key"],
		})
		expect(result.output).toMatchSnapshot()
	})

	test("reports error for different data-key with computed value", async () => {
		const { result } = await invalid({
			code: `
        <li key={foo + bar} data-key={foo} />
        `,
			errors: ["data-key-must-match-key"],
		})
		expect(result.output).toMatchSnapshot()
	})
})
