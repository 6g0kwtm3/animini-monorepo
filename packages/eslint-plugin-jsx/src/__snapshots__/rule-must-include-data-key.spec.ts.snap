// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`reports error for different data-key 1`] = `
"
        <li key={foo + bar} data-key={foo + bar} />
        "
`;

exports[`reports error for different data-key with computed value 1`] = `
"
				<li key="foo" data-key="foo" />
        "
`;

exports[`reports error for different data-key with computed value 2`] = `
"
        <li key={foo + bar} data-key={foo + bar} />
        "
`;

exports[`reports error for missing data-key 1`] = `
"
        <li key="foo" data-key="foo" />
        "
`;

exports[`reports error for missing data-key with boolean value 1`] = `
"
        <li key data-key />
        "
`;

exports[`reports error for missing data-key with computed value 1`] = `
"
        <li key={foo + bar} data-key={foo + bar} />
        "
`;

exports[`reports error when data-key is a boolean attribute without value 1`] = `
"
				<li key="foo" data-key="foo" />
        "
`;
